### PR TITLE
Proper fix for non-receiving watchers

### PR DESCRIPTION
**Pull request by wojtek-t**
Follow up on #37195

### CODE REVIEW

**Comment by k8s-oncall** [comment]
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37677)

**Comment by wojtek-t** [comment]
@k8s-bot cri node test this, issue: #IGNORE (error in queue)

**Comment by wojtek-t** [comment]
@k8s-bot verify test this, issue: #IGNORE (error starting build)

**Comment by wojtek-t** [comment]
@k8s-bot cri test this, issue: #IGNORE (error in queue)

**Comment by wojtek-t** [comment]
@k8s-bot cri test this, issue: #IGNORE (Error in queue)

**Comment by wojtek-t** [comment]
@k8s-bot etcd3 test this, issue: #IGNORE (Waiting for status to be reported)

**Comment by wojtek-t** [comment]
@k8s-bot cri node e2e test this, issue: #IGNORE (Error in queue)

File: pkg/storage/cacher.go [hunk 1]
```diff
@@ -622,6 +621,53 @@ func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
 	return strconv.ParseUint(resourceVersion, 10, 64)
 }
 
+const (
+	timeoutStep  = 50 * time.Millisecond
+	timeoutBurst = 200 * time.Millisecond
+)
+
+type selfRefreshingTimeout struct {
```
Comment (gmarek) [inline pkg/storage/cacher.go]: Please extract this struct to separate file.

File: pkg/storage/cacher.go [hunk 2]
```diff
@@ -622,6 +621,53 @@ func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
 	return strconv.ParseUint(resourceVersion, 10, 64)
 }
 
+const (
+	timeoutStep  = 50 * time.Millisecond
+	timeoutBurst = 200 * time.Millisecond
```
Comment (gmarek) [inline pkg/storage/cacher.go]: timeoutBudget?

File: pkg/storage/cacher.go [hunk 3]
```diff
@@ -622,6 +621,53 @@ func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
 	return strconv.ParseUint(resourceVersion, 10, 64)
 }
 
+const (
+	timeoutStep  = 50 * time.Millisecond
```
Comment (gmarek) [inline pkg/storage/cacher.go]: timeoutRefreshPerSecond?

File: pkg/storage/cacher.go [hunk 4]
```diff
@@ -622,6 +621,53 @@ func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
...
+				s.timeout = timeoutBurst
+			}
+			s.Unlock()
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+func (s *selfRefreshingTimeout) get() time.Duration {
```
Comment (gmarek) [inline pkg/storage/cacher.go]: I'd use 'peek' or 'read' to suggest it's not supposed to be used.

File: pkg/storage/cacher.go [hunk 5]
```diff
@@ -622,6 +621,53 @@ func (c *Cacher) LastSyncResourceVersion() (uint64, error) {
...
+	s.Lock()
+	defer s.Unlock()
+	return s.timeout
+}
+
+func (s *selfRefreshingTimeout) use(used time.Duration) {
```
Comment (gmarek) [inline pkg/storage/cacher.go]: For testing - return false when timeout became negative.

File: pkg/storage/cacher.go (hunk 1 above)
Comment (gmarek) [inline pkg/storage/cacher.go]: And please add tests for this struct.

File: pkg/storage/cacher.go [hunk 6]
```diff
@@ -735,7 +781,7 @@ func (c *cacheWatcher) stop() {
 
 var timerPool sync.Pool
 
-func (c *cacheWatcher) add(event *watchCacheEvent, timeout *time.Duration) {
+func (c *cacheWatcher) add(event *watchCacheEvent, getTimeout func() time.Duration, useTimeout func(time.Duration)) {
```
Comment (gmarek) [inline pkg/storage/cacher.go]: getAvailableTime?

File: pkg/storage/cacher.go (hunk 1 above)
Comment (gmarek) [inline pkg/storage/cacher.go]: Plus comment on the semantics (especially behavior in case of multithreading)

File: pkg/storage/cacher.go (hunk 3 above)
Comment (wojtek-t) [inline pkg/storage/cacher.go]: done

File: pkg/storage/cacher.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/storage/cacher.go]: done

File: pkg/storage/cacher.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/storage/cacher.go]: done

File: pkg/storage/cacher.go (hunk 4 above)
Comment (wojtek-t) [inline pkg/storage/cacher.go]: renamed most of things.

File: pkg/storage/cacher.go (hunk 6 above)
Comment (wojtek-t) [inline pkg/storage/cacher.go]: renamed most things.

File: pkg/storage/cacher.go (hunk 5 above)
Comment (wojtek-t) [inline pkg/storage/cacher.go]: this wasn't needed.

**Comment by wojtek-t** [comment]
@gmarek  - PTAL

**Comment by gmarek** [comment]
/lgtm

---

Reviewed 3 of 4 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/37677#-:-KYXs6BdDdAHvPxIZS6l:b-8ks5qa)*

**Comment by wojtek-t** [comment]
@k8s-bot unit test this, issue: #38518

**Pull request merged**